mender-update/update_module/v3/platform/boost_filesystem/discover.cpp

```diff
@@ -0,0 +1,79 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <mender-update/update_module/v3/update_module.hpp>
+
+#include <cerrno>
+#include <boost/filesystem.hpp>
+#include <common/conf.hpp>
+#include <common/log.hpp>
+
+// TODO: REMOVE
+#include <iostream>
+
+namespace mender {
+namespace update {
+namespace update_module {
+namespace v3 {
+
+using namespace std;
+
+namespace error = mender::common::error;
+namespace expected = mender::common::expected;
+namespace conf = mender::common::conf;
+namespace log = mender::common::log;
+namespace fs = boost::filesystem;
+
+ExpectedStringVector DiscoverUpdateModules(const conf::MenderConfig &config) {
+	vector<string> ret {};
+	fs::path dir_path = fs::path(config.data_store_dir) / "modules/v3";
+
+	try {
+		fs::directory_iterator end_itr; // default construction yields past-the-end
+		for (fs::directory_iterator itr(dir_path); itr != end_itr; ++itr) {
+			fs::path file_path = itr->path();
+			if (!fs::is_regular_file(file_path)) {
+				continue;
+			}
+
+			string file_path_str = file_path.string();
+			fs::perms perms = itr->status().permissions();
+			if ((perms & (fs::perms::owner_exe | fs::perms::group_exe | fs::perms::others_exe))
+				== fs::perms::no_perms) {
+				log::Warning("'" + file_path_str + "' is not executable");
+				continue;
+			}
+
+			// TODO: should check access(X_OK)?
```

Comment:
Q: Why not `access(X_OK` above?

Reply:
Because the ticket says it should return executable files. I'd personally do the `access()` check later.